### PR TITLE
Add microversion utilities

### PR DESCRIPTION
**Pull request by lentzi90**
This adds GetSupportedMicroversions and MicroversionSupported utility functions for working with microversions.
The doc string for ChooseVersion is also updated to clarify how it works.

Prior to a PR being reviewed, there needs to be a Github issue that the PR
addresses. Replace the brackets and text below with that issue number.

Fixes #2791

Links to the line numbers/files in the OpenStack source code that support the
code in this PR:

TODO. I am not sure what is needed here

### CODE REVIEW

File: openstack/utils/choose_version.go
Comment (dtantsur) [inline openstack/utils/choose_version.go:118]: (Thinking aloud) maybe we could return a wrapper structure with something like
```go
type SupportedMicroversions struct {
    MaxMajor int
    MaxMinor int
    MinMajor int
    MinMinor int
}
```

Then some of these functions become methods on it, like
```go
func (vs SupportedMicroversions) Require(version string) error
func (vs SupportedMicroversions) IsSupported(version string) bool
func (vs SupportedMicroversions) Choose(versions []string) string
```

etc

Comment (dtantsur) [reply openstack/utils/choose_version.go:118]: Also: you want to pass the client as a pointer, not by value.

File: openstack/utils/choose_version.go
Comment (dtantsur) [inline openstack/utils/choose_version.go:130]: When I try this, I get a segfault
```
panic: assignment to entry in nil map

goroutine 1 [running]:
github.com/gophercloud/gophercloud.(*ServiceClient).Request(0xc00010fa90, {0x6efecb, 0x3}, {0xc00001e100, 0x1a}, 0xc?)
        /home/dtantsur/gophercloud/service_client.go:150 +0x147
github.com/gophercloud/gophercloud/openstack/utils.GetSupportedMicroversions({0xc00013c000, {0xc00001e100, 0x1a}, {0x0, 0x0}, {0x6f117e, 0x9}, {0x0, 0x0}, 0xc00002ac90})
        /home/dtantsur/gophercloud/openstack/utils/choose_version.go:130 +0xd1
main.main()
        /home/dtantsur/gophercloud/main/main.go:26 +0x98
```

File: openstack/utils/choose_version.go
Comment (dulek) [inline openstack/utils/choose_version.go:24]: You're sure this is limited to "identity service", i.e. Keystone?

File: openstack/utils/choose_version.go
Comment (dulek) [inline openstack/utils/choose_version.go:179]: This should be `return false, nil`.

**Review by dulek — CHANGES_REQUESTED** [review]

File: openstack/utils/choose_version.go
Comment (dtantsur) [reply openstack/utils/choose_version.go:24]: It has a reference to `IdentityBase` down the code, dunno why.

Comment (lentzi90) [reply openstack/utils/choose_version.go:24]: Yes, I got fooled by the doc string into believing this could be used for Nova. I was very confused about why it never could find a good version to use until @mdbooth pointed out that it is taking the ProviderClient as input, not the ServiceClient. It is indeed only checking the identity endpoint and picks a version for Keystone.

File: openstack/utils/choose_version.go
Comment (lentzi90) [reply openstack/utils/choose_version.go]: Would the idea be to hold on to this wrapper and pass it around? I'm worried it would become hard to use. If I need to check/require microversions in many different places in a code base it would probably be cumbersome. Alternatively, I would end up calling `GetSupportedMicroversions` all over the place.

Comment (dtantsur) [reply openstack/utils/choose_version.go]: I only thing it will be more convenient to use methods on the wrapper rather than passing two strings around and parsing them each time. There is also a risk of people using the wrong order (what goes first, min or max)?

Comment (lentzi90) [reply openstack/utils/choose_version.go]: Ah good point! I was still in my head thinking about how I did it in CAPO, where the min and max is stored in the ComputeClient and then these functions become methods on that. Maybe I can find some way to integrate this better with the ServiceClient :thinking:

File: openstack/utils/choose_version.go
Comment (dtantsur) [reply openstack/utils/choose_version.go:24]: Given that gophercloud is working on a major release, maybe it's time to change that?

Comment (lentzi90) [reply openstack/utils/choose_version.go:24]: Yeah that would be nice!

File: openstack/utils/choose_version.go
Comment (dtantsur) [reply openstack/utils/choose_version.go:130]: https://github.com/gophercloud/gophercloud/pull/2799

Comment (dtantsur) [reply openstack/utils/choose_version.go:130]: You can avoid it by using Get instead of Request for now.

File: openstack/utils/choose_version.go
Comment (mdbooth) [reply openstack/utils/choose_version.go:24]: It would also be really good to have this in the v1 branch though, if possible. We have a bunch of potential near term use cases for it, one of which is described in https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1715#issuecomment-1752923531.

File: openstack/utils/choose_version.go
Comment (mdbooth) [inline openstack/utils/choose_version.go:144]: Would it be possible to change this API to return a copy of the ServiceClient rather than modifying the existing one? The reason for this is because I would like to ensure we always use the same 'base' microversion, and bump it explicitly only for the scope of code sections which require it. This is much simpler to do if you get a copy of the ServiceClient, because you don't have to worry about undoing it under all circumstances when you exist the relevant scope. Something like:

```go
hostnameServiceClient, err := WithMicroversion(serviceClient, "2.90")
if err != nil {
  if IsMicroversionNotSupported(err) {
    ...plan B...
  } else {
    return err
  }
}

// hostnameServiceClient has microversion 2.90
// serviceClient still has its original microversion, so no accidental microversion leaks
```

...or whatever. I'm not so bothered about the actual API so much as the principal of leaving the original unmodified so we don't have to worry about leaks.

File: openstack/utils/choose_version.go
Comment (lentzi90) [reply openstack/utils/choose_version.go]: Fixed

File: openstack/utils/choose_version.go
Comment (lentzi90) [reply openstack/utils/choose_version.go:24]: @mdbooth I thought you did not want this kind of negotiation of the version? I have reworked the CAPO PR for that reason to set an explicit version in all cases instead of using a function like this. In which cases do you think we should use this?

File: openstack/utils/choose_version.go
Comment (lentzi90) [reply openstack/utils/choose_version.go]: Done :slightly_smiling_face:

File: openstack/utils/choose_version.go
Comment (dtantsur) [inline openstack/utils/choose_version.go:179]: Why not make this a method on SupportedMicroversions?

File: openstack/utils/choose_version.go
Comment (dtantsur) [inline openstack/utils/choose_version.go:134]: This assumes a version endpoint, like `http://example.com/compute/v2`. It fails for me locally with Ironic since I have an unversioned one, which looks like this:

```json
{
  "name": "OpenStack Ironic API",
  "description": "Ironic is an OpenStack project which enables the provision and management of baremetal machines.",
  "default_version": {
    "id": "v1",
    "links": [
      {
        "href": "http://192.168.122.1:6385/v1/",
        "rel": "self"
      }
    ],
    "status": "CURRENT",
    "min_version": "1.1",
    "version": "1.87"
  },
  "versions": [
    {
      "id": "v1",
      "links": [
        {
          "href": "http://192.168.122.1:6385/v1/",
          "rel": "self"
        }
      ],
      "status": "CURRENT",
      "min_version": "1.1",
      "version": "1.87"
    }
  ]
}
```

Comment (dtantsur) [reply openstack/utils/choose_version.go:134]: Maybe we can check the `Versions` list, and if it contains only one item, assume it's the correct one? Otherwise, fail with a clear error message?

File: openstack/utils/testing/choose_version_test.go
Comment (dtantsur) [inline openstack/utils/testing/choose_version_test.go:201]: Could you test for a reasonable error message? With Ironic I get `invalid microversion format: ""` which is quite confusing.

File: openstack/utils/choose_version.go
Comment (mdbooth) [reply openstack/utils/choose_version.go:24]: I probably got confused and thought I was commenting on the PR in general :)

You are correct, I don't want to do negotiation in CAPO: we either get a specific version or we don't execute the code. There may be cases where some code might run with one of a small number of specific microversions (e.g. GET server) but we should never, e.g. request microversion N and get microversion N+1.

That said, it might be useful to be able to specify a microversion limit for testing so we can test old microversions against a newer server. The microversion requested would still be at the point of use, though, and we would still either get it or not get it.

However, CAPO hopefully isn't the only consumer of this code. Negotiation might be useful elsewhere, e.g. a CLI utility.

File: openstack/utils/choose_version.go
Comment (lentzi90) [reply openstack/utils/choose_version.go]: Done (and renamed to `IsSupported`)

File: openstack/utils/choose_version.go
Comment (lentzi90) [reply openstack/utils/choose_version.go]: I have updated with something I believe does what you suggested. I even added a test with a made up Ironic endpoint that returns multiple versions to check the error case. Not sure if it makes sense :sweat_smile:

**Review by dtantsur — APPROVED** [review]
Works with Ironic in my testing, thank you!

**Comment by lentzi90** [comment]
Thank you for testing!
I have also confirmed that the code works with CAPO in https://github.com/kubernetes-sigs/cluster-api-provider-openstack/pull/1567 (although I had to backport it to the v1 branch first since that is what CAPO is using)

**Review by EmilienM — APPROVED** [review]
Nice work, I'll wait for others to approve or request changes.

**Comment by pierreprinetti** [comment]
I want to look at this this afternoon; please go ahead and merge if I don’t

**Comment by pierreprinetti** [comment]
> I want to look at this this afternoon; please go ahead and merge if I don’t

I still believe I can get to that today :crossed_fingers:

File: openstack/utils/choose_version.go
Comment (pierreprinetti) [inline openstack/utils/choose_version.go:136]: I wonder if we can assume that all services will either use `version`, `versions`, or no versions. What if we have service-specific logic instead? Have you considered having service-specific logic, maybe exposed in the ServiceClient through an interface?

File: openstack/utils/choose_version.go
Comment (pierreprinetti) [inline openstack/utils/choose_version.go:181]: I like this function signature. You give me a version, and I tell you if it's OK. This requires many less assumptions. Can we make it so that this is the only new exported symbol, and unexport all the rest?

**Review by pierreprinetti — CHANGES_REQUESTED** [review]

**Comment by pierreprinetti** [comment]
Thank you for this addition. I think that the functionality is useful.

File: openstack/utils/choose_version.go
Comment (dtantsur) [reply openstack/utils/choose_version.go:122]: "the fact that they're split into exactly major and minor" is the hard fact that is embedded into the definition of microversions. I'm fine with hiding private details as long as I can still access them via accessors (not just the string representations - actual numbers).

File: openstack/utils/choose_version.go
Comment (dtantsur) [reply openstack/utils/choose_version.go:125]: Your example is too hypothetical :) First, it assumes having two major versions on one endpoint, which is not how things tend to be done. Second, it assumes two major versions *with microversions* which is a thing that is quite unlikely to happen because the very idea of microversions implies that you can do breaking changes without creating major versions. We do ponder Ironic v2 from time to time, but that's more of a wishful thinking (file it under "rewrite Ironic in Go").

File: openstack/utils/choose_version.go
Comment (dtantsur) [reply openstack/utils/choose_version.go:181]: Please don't. GopherCloud is already bare bones enough to justify another high-level client built upon it.

File: openstack/utils/choose_version.go
Comment (dtantsur) [reply openstack/utils/choose_version.go:136]: In theory - no. In practice, everyone just copied what Nova did :)

Comment (dtantsur) [reply openstack/utils/choose_version.go:136]: As a data point, I've implemented a similar thing in rust-osauth long ago, and nobody complained so far: https://github.com/dtantsur/rust-osauth/blob/master/src/protocol.rs#L31-L39

File: openstack/utils/choose_version.go
Comment (dtantsur) [reply openstack/utils/choose_version.go:122]: Reference: https://specs.openstack.org/openstack/api-sig/guidelines/microversion_specification.html

File: openstack/utils/choose_version.go
Comment (pierreprinetti) [reply openstack/utils/choose_version.go:125]: The way I see it, the assumption is to think that my example won’t happen. Which may or may not be a safe assumption.

File: openstack/utils/choose_version.go
Comment (pierreprinetti) [reply openstack/utils/choose_version.go:136]: How is Swift versioning its API?

File: openstack/utils/choose_version.go
Comment (pierreprinetti) [reply openstack/utils/choose_version.go:181]: I may be failing to see your intended use case. Do you have one in mind?

The use case I’m most inclined to think: a client has two max three behaviours coded for a specific action, each corresponding to a specific API. The client would ask for its preferred version, then if not available it would test the fallback behaviour, and so on. In this case the client knows what exact microversion it needs and asks Goohercloud whether Itshak OK. Do you have a case where a client needs something different, like a range?

Comment (dtantsur) [reply openstack/utils/choose_version.go:181]: Largely, I want to know *in advance* what features I can and cannot use, so that I don't need to delay this decision until much later in the code. Our code is organized in the way that the controller layer does not know about the Ironic implementation details, and we'd rather keep it that way.

File: openstack/utils/choose_version.go
Comment (dtantsur) [reply openstack/utils/choose_version.go:136]: I don't think it actually does. Of the core services, I think only Nova, Keystone and Cinder (and Ironic if you count it core) use microversioning.

File: openstack/utils/choose_version.go
Comment (dtantsur) [reply openstack/utils/choose_version.go:125]: It's a question of whether you want to try covering 100% of potential future cases. It's not a terrible idea if you do, but that also mean you will do much more work than you need to. With rust-openstack, I took the approach of "I'll think about it when it happens" (fwiw, openstacksdk does the same), which has its own pros and cons.

**Comment by pierreprinetti** [comment]
I will not have the opportunity to come back to this PR in the next week; @EmilienM please go ahead and do as you see fit

File: openstack/utils/choose_version.go
Comment (lentzi90) [reply openstack/utils/choose_version.go:125]: From what I can see, this (min and max) is the only information we can get from a single endpoint so there is nothing more we can do.

**Comment by lentzi90** [comment]
> A more general comment: the most Gophercloud-y solution for this problem would probably be to have a `versions` package in each relevant service, with a `List` function that makes the call to the service and returns a typed response. The caller would then adjust their ServiceClient microversion accordingly, with or without a helper function in Gophercloud.
> 
> I wonder if this "generic" solution that aims at encapsulating the calls and solving negotiation in one swipe, isn't going to feel like a trap going forward.

Thank you for the comment! I appreciate the insights and would ask that the maintainers decide on what would be the preferred way forward here. I can see pros and cons witch both approaches.

I want to also highlight this function in the utils repo: https://github.com/gophercloud/utils/blob/80377eca5d56d1f07d0ebdd4c2a5ae000a66a90e/openstack/clientconfig/requests.go#L739
Would it make more sense to move the microversion utilities there? It also handles all services in one function so perhaps it would fit in better there.

**Comment by dtantsur** [comment]
> the most Gophercloud-y solution for this problem would probably be to have a versions package in each relevant service, with a List function that makes the call to the service and returns a typed response

Could be, but currently Gophercloud is handling Microversions as a standard feature (which they are in a sense), e.g. they handling is hardcoded in the ServiceClient: https://github.com/gophercloud/gophercloud/blob/69b57d9428c814942f444f049b6d6a1aebd82600/service_client.go#L116. If we start treating Microversions as service features, we need to change that as well. Which is probably a lot of work and will make the change no longer backportable..

**Comment by dtantsur** [comment]
@lentzi90 @EmilienM @pierreprinetti so folks, what's the path forward here? I can leave with a reduced scope if that allows this change to move forward.

**Comment by EmilienM** [comment]
I feel like this is a great iteration moving forward.

From Pierre:
> I wonder if this "generic" solution that aims at encapsulating the calls and solving negotiation in one swipe, isn't going to feel like a trap going forward.

I agree but I also think this would be doable in a future version. I want to unblock these folks for now.

**Pull request merged**
